doc/source/design_manifesto.rst

```diff
@@ -0,0 +1,131 @@
+==================
+ Design Manifesto
+==================
+
+The Nouns
+=========
+
+
+``Header``
+----------
+ - Attributes
+    - ``Start`` Document
+    - ``Stop`` Document
+        - may want to flip non-existant behavior from ``{}`` -> ``None``
+    - Reference back to a ``DataBroker``
+
+ - properties (ex, LAZY):
+   - ``streams`` list of streams from all of the event stores
+   - ``descriptors`` mapping to lists of descriptors per stream
+
+ - Methods:
+   - ``h.stream(stream_name, fill=True) -> generator``
+     - generator for just this stream,
+     - (name, doc) for all document types
+   - ``h.table(stream_name, fill=True) -> DataFrame``
+     - Dataframe for just this stream
+   - ``h.events(fill=True) -> generator``
+     - all documents as (name, doc) pair
+     - this should maybe be renamed?
+   - ``h.es_given_stream(stream_name) -> EventSource``
+  - mimic ``doct.Document`` interface?
+
+
+The implementation of these methods should live in the ``Header`` class using the resources
+provided by the attached ``DataBroker``
+
+
+``DataBroker``
+--------------
+ - Attributes:
+   - Exactly 1 ``HeaderSource``
+   - ``EventSource`` list
+ - Methods:
+  - mirror out all search from ``HeaderSource``, converts to ``Header`` objects
+  - manage filters + search state
+  - drop everything else!
+  - insert?
+    - how to handle ES distribution.
+
+``HeaderSource``
+----------------
+ - databases, not public:
+   - ``Start`` document collection / table
+   - ``Stop`` document collection / table
+
+ - Methods
+   - provides search capabilities over the start / stop documents
+   - ``hs[uid] -> (Start, Stop)``
+   - ``hs[scan_id] -> (Start, Stop)``
+   - ``hs[-offset] -> (Start, Stop)``
+   - ``hs(full=text, search=method) -> (Start, Stop)``
+
+``EventSource``
+---------------
+ - databases, not public:
+   - ``Event`` collection
+   - ``Descriptor`` document collection
+   - ``Filestore`` collection(s) / object
+
+ - Methods:
+   - insert ``es.insert(name, doc)``
+   - get streams given a ``Header``
+     - ``es.streams_given_header(header) -> {stream_names}``
+
+   - get descriptors
+     - ``es.descriptors_given_header(header, **kwargs) -> [Descriptor]``
+     - ``es.descriptor_given_stream(header, stream, **kwargs) -> Descriptor``
```

Comment:
I think this can probably be one method  (`descriptors()`) with an optional `stream` argument.